### PR TITLE
New Session History

### PR DESCRIPTION
**Pull request by cbrewster**
Remaining Work:
 - [x] Move `LoadData` from `BrowsingContext` to `Pipeline`
 - [x] Cleanup `*Diff` and `*Changeset` types
 - [x] Implement pipeline discarding and reloading
 - [x] Document new code

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20507)

### CODE REVIEW

**Comment by gterzian** [comment]
Hi @cbrewster 

I have a question:

In the context of #20329, I have to unload the 'active document' of the browsing context of the entry to which we are traversing, if the document of the entry is not the currently active one in the target browsing context. (see https://html.spec.whatwg.org/multipage/history.html#traverse-the-history-by-a-delta Step 5.2)

I'm currently doing this there: https://github.com/servo/servo/pull/20329/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fR2041

It's a bit of a hack, I'm wondering if it's the best way, and if there is a better one, perhaps you could make it available as part of this PR?

Also, am I correct to think that the concept of the 'current entry' of a browsing context, is same as the 'active document' that the spec refers to? The spec refers to "A browsing context's active document is its WindowProxy object's [[Window]] internal slot value's associated Document.", which in Servo is a pipeline id, I assumed it's the same pipeline as the 'current entry, but wasn't sure.

Thanks.

**Comment by asajeffrey** [comment]
OK, read through the current WIP, it looks much much nicer! Yay for not having to keep those timestamps around!

A couple of questions...

a) Way back when, we were keeping the session history as a list of pipeline ids, then we ripped it out and replaced it with the current solution of timestamps and explicit iterators. I'm not sure why we didn't go for this implementation then? Did we just not think of it?

b) In `session_history.rs` there's `SessionHistoryChange`, `BrowsingContextDiff` and `SessionHistoryDiff`, all of which do very similar jobs. I wonder if there's a way to unify all of these into one structure?

I also need to think about how we're handling unloading documents, and when the session history gets updated. Also, it might be worth writing an addendum to the navigation history paper talking about this implementation strategy.

**Comment by asajeffrey** [comment]
Perhaps we can unify all the `*Diff` types into one, and maybe handle the history state changes by allowing a diff where the old and new pipeline ids are the same?

I think `SessionHistoryChange` will have to stay, because its old pipeline id is optional, and is `None` in the case of the first session history entry.

**Comment by cbrewster** [comment]
My concern with combining the `*Diff` types into 1 is that the `BrowsingContextDiff` and `PipelineDiff` hold very different data. The former holds info about pipeline ids and load data for a specific browsing context and the latter holds info about hash changes/state changes for a specific pipeline.

As for `SessionHistoryChange`, I am wondering if we could get rid of `old_pipeline_id`. It is only used here:
https://github.com/servo/servo/blob/122bfa03e1fba01dc7d52e63268a83492d994b53/components/constellation/constellation.rs#L1684-L1688
We could get the same pipeline by getting the active pipeline_id from the browsing context.

**Comment by asajeffrey** [comment]
Hmm, I see what you mean about the `*Diff`s. Perhaps a `SessionHistoryDiff` with something like:
```rust
enum SessionHistoryDiff {
  Initial {
    pipeline_id: PipelineId, 
    browsing_context_id: BrowsingContextId,
    load_data: LoadData,
  },
  Navigate {
    old_pipeline_id: PipelineId, 
    new_pipeline_id: PipelineId, 
    browsing_context_id: BrowsingContextId,
    old_load_data: LoadData,
    new_load_data: LoadData,
  },
  PushState {
    pipeline_id: PipelineId,
    browsing_context_id: BrowsingContextId,
    old_history_state: HistoryStateId,
    new_history_state: HistoryStateId,
  },
}
```
I think the reason why we're keeping the old pipeline id in the change is so we can discard the change if the browsing context has already changed? E.g. a race between two navigations?

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -2470,7 +2464,7 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
                 .filter_map(|maybe_pipeline| maybe_pipeline)
                 .collect::<Vec<_>>();
 
-            pipelines_to_evict.extend(session_history.future.iter()
+            pipelines_to_evict.extend(session_history.future.iter().rev()
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Really? The session history future is in increasing chronological order isn't it? We should be evicting ones far off in the future, not the near future.

File: components/constellation/browsingcontext.rs [hunk 2]
```diff
@@ -29,72 +25,33 @@ pub struct BrowsingContext {
     /// The size of the frame.
     pub size: Option<TypedSize2D<f32, CSSPixel>>,
 
-    /// The timestamp for the current session history entry.
-    pub instant: Instant,
```
Comment (asajeffrey) [inline components/constellation/browsingcontext.rs]: Yay! No more timestamps!

File: components/constellation/constellation.rs [hunk 3]
```diff
@@ -277,6 +274,8 @@ pub struct Constellation<Message, LTF, STF> {
     /// to become same-origin, at which point they can share DOM objects.
     event_loops: HashMap<TopLevelBrowsingContextId, HashMap<Host, Weak<EventLoop>>>,
 
+    joint_session_histories: HashMap<TopLevelBrowsingContextId, SessionHistory>,
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Bikeshed: should we rename `SessionHistory` to `TopLevelBrowsingContext`?

File: components/constellation/constellation.rs [hunk 4]
```diff
@@ -1732,14 +1661,17 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         let IFrameLoadInfo {
             parent_pipeline_id,
             new_pipeline_id,
-            replace,
+            replace: _replace,
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Use a `...` pattern for this?

File: components/constellation/constellation.rs [hunk 5]
```diff
@@ -1995,45 +1919,139 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+                    if session_history.future.len() < forward {
+                        return warn!("Cannot traverse that far into the future.");
+                    }
+
+                    for _ in 0..forward {
+                        let diff = session_history.future.pop().expect("Infallible");
+                        match diff {
+                            SessionHistoryDiff::BrowsingContextDiff {
+                                browsing_context_id,
+                                ref new_pipeline_id,
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Do we need this `ref`? Does `PipelineId` not impl `Copy`?

File: components/constellation/constellation.rs (hunk 1 above)
Comment (cbrewster) [reply components/constellation/constellation.rs]: Yes we want to reverse here because this will put the diffs in the near future first, which we then skip by `history_length`.

File: components/constellation/constellation.rs [hunk 6]
```diff
@@ -1995,45 +1919,139 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+                .entry(top_level_browsing_context_id).or_insert(SessionHistory::new());
+
+            match direction {
+                TraversalDirection::Forward(forward) => {
+                    if session_history.future.len() < forward {
+                        return warn!("Cannot traverse that far into the future.");
+                    }
+
+                    for _ in 0..forward {
+                        let diff = session_history.future.pop().expect("Infallible");
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Can we remove this source of panic?

File: components/constellation/constellation.rs [hunk 7]
```diff
@@ -2474,53 +2384,111 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+    }
+
+    fn trim_history(&mut self, top_level_browsing_context_id: TopLevelBrowsingContextId) {
+        let pipelines_to_evict = {
+            let session_history = self.joint_session_histories.entry(top_level_browsing_context_id)
+                .or_insert(SessionHistory::new());
+
+            let history_length = PREFS.get("session-history.max-length").as_u64().unwrap_or(20) as usize;
+
+            let mut pipelines_to_evict = session_history.past.iter().rev()
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Add a comment about the use of `rev()` here (as in, I can't remember which order we get the session history past in, so this would help future me).

File: components/constellation/pipeline.rs [hunk 8]
```diff
@@ -89,6 +89,9 @@ pub struct Pipeline {
     /// Whether this pipeline should be treated as visible for the purposes of scheduling and
     /// resource management.
     pub visible: bool,
+
+    /// The Load Data used to create this pipeline.
+    pub load_data: LoadData,
```
Comment (asajeffrey) [inline components/constellation/pipeline.rs]: Yay!

File: components/constellation/session_history.rs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+use msg::constellation_msg::{BrowsingContextId, PipelineId, TopLevelBrowsingContextId};
+use script_traits::LoadData;
+use std::{fmt, mem};
+use std::cmp::PartialEq;
+
+/// Represents the joint session history
+/// https://html.spec.whatwg.org/multipage/#joint-session-history
+#[derive(Debug)]
+pub struct SessionHistory {
+    /// Diffs used to traverse to past entries
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: Add a comment re sort order.

File: components/constellation/session_history.rs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    pub past: Vec<SessionHistoryDiff>,
+
+    /// Diffs used to traverse to future entries
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: Ditto.

File: components/constellation/session_history.rs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    /// The pipeline for the document being loaded.
+    pub new_pipeline_id: PipelineId,
+
+    /// The old pipeline that the new pipeline should replace.
+    pub replace: Option<AliveOrDeadPipeline>,
+}
+
+/// Represents a pipeline or discarded pipeline in a history entry.
+#[derive(Clone, Debug)]
+pub enum AliveOrDeadPipeline {
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: Bikeshed the name. Is there a spec concept we can use? Link to?

File: components/constellation/session_history.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    /// Represents a pipeline that has not been discarded
+    Alive(PipelineId),
+    /// Represents a pipeline that has been discarded and must be reloaded with the given `LoadData`
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: Nit: "must be reloaded with the given `LoadData` if it is ever traversed to".

File: components/constellation/session_history.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+}
+
+impl PartialEq for AliveOrDeadPipeline {
+    fn eq(&self, other: &AliveOrDeadPipeline) -> bool {
+        match *self {
+            AliveOrDeadPipeline::Alive(pipeline_id) => {
+                match *other {
+                    AliveOrDeadPipeline::Alive(other_pipeline_id) if pipeline_id == other_pipeline_id => true,
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: `AliveOrDeadPipeline::Alive(other_pipeline_id) => (pipeline_id == other_pipeline_id),`

File: components/constellation/session_history.rs [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+                    _ => false,
+                }
+            },
+            AliveOrDeadPipeline::Dead(pipeline_id, _) => {
+                match *other {
+                    AliveOrDeadPipeline::Dead(other_pipeline_id, _) if pipeline_id == other_pipeline_id => true,
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: Ditto.

File: components/constellation/session_history.rs [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    pub fn alive_pipeline_id(&self) -> Option<PipelineId> {
+        match *self {
+            AliveOrDeadPipeline::Alive(pipeline_id) => Some(pipeline_id),
+            AliveOrDeadPipeline::Dead(..) => None,
+        }
+    }
+}
+
+impl PartialEq for AliveOrDeadPipeline {
+    fn eq(&self, other: &AliveOrDeadPipeline) -> bool {
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: Add a comment about why we're not just using the derived version.

File: components/constellation/session_history.rs [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+            }
+        }
+    }
+
+    /// Replaces all occurances of the replaced pipeline with a new pipeline
+    pub fn replace(&mut self, replaced_pipeline_id: &AliveOrDeadPipeline, pipeline_id: &AliveOrDeadPipeline) {
+        match *self {
+            SessionHistoryDiff::BrowsingContextDiff { ref mut old_pipeline_id, ref mut new_pipeline_id, .. } => {
+                if *old_pipeline_id == *replaced_pipeline_id {
+                    *old_pipeline_id = pipeline_id.clone();
```
Comment (asajeffrey) [inline components/constellation/session_history.rs]: can we not copy the pipeline ids?

File: ports/servo/browser.rs [hunk 17]
```diff
@@ -122,14 +122,18 @@ impl Browser {
             }
             (CMD_OR_ALT, None, Key::Right) | (KeyModifiers::NONE, None, Key::NavigateForward) => {
                 if let Some(id) = self.browser_id {
-                    let event = WindowEvent::Navigation(id, TraversalDirection::Forward(1));
-                    self.event_queue.push(event);
+                    if state == KeyState::Pressed {
```
Comment (asajeffrey) [inline ports/servo/browser.rs]: This looks like it should be in a separate PR.

File: ports/servo/browser.rs [hunk 18]
```diff
@@ -122,14 +122,18 @@ impl Browser {
...
+                        self.event_queue.push(event);
+                    }
                 }
             }
             (CMD_OR_ALT, None, Key::Left) | (KeyModifiers::NONE, None, Key::NavigateBackward) => {
                 if let Some(id) = self.browser_id {
-                    let event = WindowEvent::Navigation(id, TraversalDirection::Back(1));
-                    self.event_queue.push(event);
+                    if state == KeyState::Pressed {
+                        let event = WindowEvent::Navigation(id, TraversalDirection::Back(1));
```
Comment (asajeffrey) [inline ports/servo/browser.rs]: Ditto.

**Review by asajeffrey — COMMENTED** [review]
Yay, this is much nicer!

**Comment by asajeffrey** [comment]
Remove the WIP tag?

File: components/constellation/constellation.rs (hunk 6 above)
Comment (cbrewster) [reply components/constellation/constellation.rs]: Done.

File: components/constellation/constellation.rs (hunk 7 above)
Comment (cbrewster) [reply components/constellation/constellation.rs]: Done.

File: components/constellation/session_history.rs (hunk 9 above)
Comment (cbrewster) [reply components/constellation/session_history.rs]: Done.

File: components/constellation/session_history.rs (hunk 10 above)
Comment (cbrewster) [reply components/constellation/session_history.rs]: Done.

File: components/constellation/session_history.rs (hunk 11 above)
Comment (cbrewster) [reply components/constellation/session_history.rs]: IRC: `NeedsToReload` with `Yes` and `No` variants.

File: components/constellation/session_history.rs (hunk 13 above)
Comment (cbrewster) [reply components/constellation/session_history.rs]: Done.

File: components/constellation/session_history.rs (hunk 14 above)
Comment (cbrewster) [reply components/constellation/session_history.rs]: Done.

File: components/constellation/session_history.rs (hunk 12 above)
Comment (cbrewster) [reply components/constellation/session_history.rs]: Done.

File: components/constellation/constellation.rs (hunk 5 above)
Comment (asajeffrey) [reply components/constellation/constellation.rs]: Ah I see, yes a different name would help!

File: components/constellation/session_history.rs (hunk 16 above)
Comment (asajeffrey) [reply components/constellation/session_history.rs]: Variable naming?

**Comment by asajeffrey** [comment]
OK, just some minor nits left, then we can squash and merge! I will trust you to do that, and I'm on PTO tomorrow, so you can r=me.
